Improved UART handling

hfog opened the pull request:
This fixes a number of problems encountered when connecting Octoprint to the TFT over UART3 (or 4).
When Octoprint is connected over UART, the TFT can now be in either Touch mode or Emulation mode and it will still work. You can even switch between the two during printing from Octoprint (although I think there is still some risk of dropping a command - more testing is needed there).
It also passes unsolicited ack messages through to Octoprint, which enables Octoprint to stay connected when using autoreporting (ex: M105 T2).
The code is also now smart enough to not poll for temps if it's already passing temp messages on to Octoprint, avoiding unnecessary traffic over the serial line.
The next step from this would be to come up with a way for the TFT to reliably detect when a print is started from Octoprint. I can't think of any reliable way to do this though without requiring Octoprint to send some additional messages to the TFT, similar to what Marlin sends during an SD print. Technically this isn't really necessary though as the current behavior works quite well and it wouldn't really add much.

This fixes a number of issues but the primary one is https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/281

hfog committed:
Fix printing from an external device connected to one of the UART ports (such as Octoprint).
The goal is for this to work properly in either touch mode or emulation mode.
This should also allow switching between touch and emulation mode even while printing.

Commit:
Strip line numbers and checksums from gcode for parsing.
BEWARE that this probably will result in errors since the TFT will likely be thinking all of these commands are coming from its own system. In order for this to work properly, I'll probably need to add logic to make some parsing contingent upon where the gcode originated from.

hfog committed:
Revert "Strip line numbers and checksums from gcode for parsing."

This reverts commit bf164af943011bc4817434565772619613fd23ae.

hfog committed:
Change strategy for parsing gcode. It now forks into one section for TFT originating gcode and one section for external originating gcode.
This allows synchronizing feed rate and fan speed.

hfog committed:
Pass spontaneous messages (messages that aren't part of an explicit ack) to all devices connected over UART. (Keeps messages like alerts, errors, and temp heartbeats from getting lost.)
Don't have the TFT send more M105 temp messages if it doesn't have a reason to. (Temps sent by request to other devices are getting parsed so there's no need to ask for more if you've just seen one.)

hfog committed:
Minor tweaks to make sure we continue to pump message loops.

hfog committed:
Merge pull request #1 from guruathwal/master

Fix boot logo to original size for TFT24/28

hfog committed:
Merge commit 'c326dd0fe8c9876053f3409c423790b7ea8f2ffc' into printfromuart

hfog commented:
Apologies for forking from your repo instead of the main one but I really wanted the new menu functionality you've been working on and this turned out to be a more stable place to build from, given the problems I was trying to solve.

hfog commented on TFT/src/User/Configuration.h:
```diff
@@ -174,7 +178,7 @@
  * Note: this might make the  graphics slow when switching menus while printing.
 */
-//#define RAPID_SERIAL_COMM
+#define RAPID_SERIAL_COMM
```
My experience is that the microcontroller in the TFT is more than fast enough to enable this without any observed UI slowness and it really ought to be enabled. Otherwise you WILL get stuttering during printing. Even with this enabled, I'm noticing some stuttering when printing in touch mode. I think there's still a lot of room for optimization.

hfog commented on TFT/src/User/Menu/Heat.c:
```diff
@@ -266,20 +266,24 @@ void menuHeat(void)
-    if(update_waiting == true)                {nowTime=OS_GetTime();break;}
-    if(OS_GetTime()<nowTime+update_time)       break;
+    if(update_waiting == true)                {updateLastHeatCheckTime();break;}
+    if(OS_GetTime() - lastHeatCheckTime < update_time)       break;
```
FYI, when comparing two unsigned values like this, it's much safer to subtract the larger from the smaller and then compare to the delta, as I've done here. That will avoid any potential problems due to accidental comparison with signed values or due to integer overflow. This way is ALWAYS guaranteed to work the way you think it will.

guruathwal commented:
Great work... I do not use Octoprint so I can not test it.
it will be nice if @gneu42 , @scope666 & @JWolfJr can test.

Also there need to be an option to enable / disable this for users who do not use Octoprint and use USB to connect the main board to PC. This may interfere when printing via PC with 12864 emulator enabled on TFT.

Scope666 commented:
> Great work... I do not use Octoprint so I can not test it.
> it will be nice if @gneu42 , @Scope666 & @JWolfJr can test.
> 
My original TFT35 was returned to Amazon because of all the Octoprint issues, BUT BTT is sending me one from China as a result of my original SKR Mini E3 having bad driver issues.  I'll definitely test but it will take awhile before it arrives.

hfog commented:
> Great work... I do not use Octoprint so I can not test it.
> it will be nice if @gneu42 , @Scope666 & @JWolfJr can test.
> 
> Also there need to be an option to enable / disable this for users who do not use Octoprint and use USB to connect the main board to PC. This may interfere when printing via PC with 12864 emulator enabled on TFT.

Agreed. If you comment out #CLEAN_MODE_SWITCHING_SUPPORT in config.h then that will disable the polling when in emulator mode. If you would rather put that behind a setting in the setting menu then be my guest.

Along similar lines, I'm fairly sure that the firmware is interacting with the serial port even with that code disabled (before my changes). See this conversation: https://github.com/bigtreetech/BTT-TFT35-E3-V3.0/issues/10

hfog commented:
And while I have your eyes, I have a couple of questions.

1. Is there any logging going on in the firmware, and if so, how does one enable/disable it and get access to the generated logs.
2. Is there any facility for instrumenting for performance data? I'd love to know which routines are using the most cycles and why.
3. What is the recommended procedure for hooking up a debugger to the TFT? There must be a way to do it. Right?

JWolfJr commented:
I'd be happy to check this out. A couple of questions though. Should I still have the octopi hooked up through the USB cable? Also use this version of Marlin and uncomment clean mode switching is all that is needed?

hfog commented:
@JWolfJr This is primarily for hooking octopi up through one of the UART ports on the TFT. The primary benefit to doing this is that the TFT becomes aware of the traffic coming from Octoprint so it can keep its locations, temps, fan settings, etc. up to date and it ensures that Marlin only sees a single serial stream, reducing possible contention on the device. The secondary benefit is that you can switch between touch mode and emulation mode and Octoprint will still work when hooked up this way.
It shouldn't make using Octoprint over USB any worse but that isn't the primary use case and you won't get the benefits if you don't hook Octoprint to the TFT.
See my comments here: https://github.com/bigtreetech/BTT-TFT35-E3-V3.0/issues/10 if you need to know how to hook your Octopi to the TFT.
And this is not a version of Marlin. This is a version of the TFT firmware. You'll have to build it with #CLEAN_MODE_SWITCHING_SUPPORT uncommented (which it is by default) and flash your TFT.

JWolfJr commented:
Ok, gotcha. I got the specs how to hook it up to the tft, from what I saw it's going to the wifi headers? I apologise, I thought I was on the MB GitHub page. I'm at work now, I will give this a try soon as possible and report back . Thank you.

hfog commented:
You could hook it through the wifi headers but honestly, it's easier to hook it to UART3 or UART4.
Any of the serial inputs on the TFT should work.

hfog commented:
Oh and BTW, You only need CLEAN_MODE_SWITCHING_SUPPORT if you want to be able to drive Octoprint through the TFT while in emulation mode. If you're only interested in driving Octoprint through the TFT in touch mode then you don't need this uncommented.

JWolfJr commented:
I'm interested in both. I'll build the updated version and try it out. Thanks for clarifying.

guruathwal commented:
@hfog 
> Agreed. If you comment out #CLEAN_MODE_SWITCHING_SUPPORT in config.h then that will disable the polling when in emulator mode. If you would rather put that behind a setting in the setting menu then be my guest.
> 

ok. sorry I failed to review the code at that time.

> Along similar lines, I'm fairly sure that the firmware is interacting with the serial port even with that code disabled (before my changes). See this conversation: [bigtreetech/BTT-TFT35-E3-V3.0#10](https://github.com/bigtreetech/BTT-TFT35-E3-V3.0/issues/10)
 
It does not look like there is any thing related to serial comm. in the Simulator Loop and in the function calls through it (this need more digging into the code):
https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/5777f41b5f8c1a41410a6874614499c91ac78fa2/TFT/src/User/API/UI/ST7920_Simulator.c#L139-L156

It may be possible to confirm this by connecting the TFT Controller to PC via USB-Serial instead the main board. One point to note is that the serial pins on the TFT Controller are being set to Input Floating State when the serial is disconnected:
https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/7a370bd427b9722ae0be797b006c1f1a30db6b00/TFT/src/User/Hal/stm32f10x/usart.c#L20-L25


> And while I have your eyes, I have a couple of questions.
> 
> 1. Is there any logging going on in the firmware, and if so, how does one enable/disable it and get access to the generated logs.
> 2. Is there any facility for instrumenting for performance data? I'd love to know which routines are using the most cycles and why.
> 3. What is the recommended procedure for hooking up a debugger to the TFT? There must be a way to do it. Right?

There is no logging/debugging code in the actual firmware and the comments are in Chinese which also got scrambled when the commits were merged due to wrong text encoding, but the libraries used are mostly from STMICROELECTRONICS. So it should be possible to debug it using the serial wire debugging (SWD) port on the device using ST-LINK. I have not gone any further into this.
Every page on the firmware has its own while loop, it should not be that hard to point out the current loop when it is on a certain page.
Loop inside loop also means that the variables in the parent function of the parent loop will also stay in the memory until the the end of the function??

hfog commented:
> One point to note is that the serial pins on the TFT Controller are being set to Input Floating State when the serial is disconnected:
> https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/7a370bd427b9722ae0be797b006c1f1a30db6b00/TFT/src/User/Hal/stm32f10x/usart.c#L20-L25

Which is why I've made the changes I've made in Mode.c. When CLEAN_MODE_SWITCHING_SUPPORT is defined, the serial resources are left initialized when you switch out of Touch mode.

gilliands commented:
Tested on the TFT35 E3 V3 with an ESP-01S and a TFT24 V1.1 with an ESP12-E and confirmed working in emulation mode. Great work.

guruathwal committed:
translate comments to english in popup.c & disable ST7920_FULLSCREEN by default

hfog reviewed (commented):
Thanks!

guruathwal committed:
move `CLEAN_MODE_SWITCHING_SUPPORT` to Marlin mode Setting area and add sanity check

move `CLEAN_MODE_SWITCHING_SUPPORT` to Marlin mode Setting area.
add sanity check  for `CLEAN_MODE_SWITCHING_SUPPORT`  to throw error if it is enabled for TFT controllers with no Emulator/Marlin mode support.

guruathwal committed:
Update Configuration.h
